### PR TITLE
Improve datetime field editing

### PR DESCRIPTION
**Pull request by nCastle1**
Improves support for editing datetime fields. New datetime editor has been designed to work well for both keyboard-driven and mouse/touch-driven data entry.

### CODE REVIEW

File: src/DataCollection.UWP/Controls/DateTimeEditor.xaml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        <Grid.RowDefinitions>
+            <RowDefinition Height="38" />
+            <RowDefinition Height="Auto" />
+            <RowDefinition Height="Auto" />
+        </Grid.RowDefinitions>
+
+        <TextBox x:Name="DatePartTextBox"
+                 GotFocus="TextBox_GotFocus"
+                 LostFocus="DatePartTextBox_TextChanged"
+                 PlaceholderText="{x:Bind Converter={StaticResource LC}, ConverterParameter=DTE_DatePart_Placeholder}"
```
Comment (ZackAllen) [inline src/DataCollection.UWP/Controls/DateTimeEditor.xaml:47]: LC could be a longer name like LocConverter. Up to you.

File: src/DataCollection.UWP/LocalizedStrings/Resources.resw [hunk 2]
```diff
@@ -540,4 +540,28 @@
...
+  <data name="DTE_ValidationError_Message" xml:space="preserve">
+    <value>Set date and time to valid values before moving on</value>
+  </data>
+  <data name="DTE_EmptyDate_Message" xml:space="preserve">
+    <value>Saved as: null (no date/time)</value>
+  </data>
+  <data name="DTE_Preview_FormatString" xml:space="preserve">
+    <value>Saved as: {0} {1} (local time)</value>
+  </data>
 </root>
```
Comment (ZackAllen) [inline src/DataCollection.UWP/LocalizedStrings/Resources.resw]:
```
  <data name="TP_CancelButton_Tooltip" xml:space="preserve">
    <value>Discard Selected Time</value>
  </data>
  <data name="TP_SaveButton_Tooltip" xml:space="preserve">
    <value>Save Selected Time</value>
  </data>
```
When I build the project, these lines are added after 566. Not sure why they arent already in your branch.

**Review by ZackAllen — APPROVED** [review]
The UI looks excellent on UWP and WPF. I built and tested on every platform. Code and xaml look very good as well.

File: src/DataCollection.UWP/LocalizedStrings/Resources.resw (hunk 2 above)
Comment (nCastle1) [reply src/DataCollection.UWP/LocalizedStrings/Resources.resw]: Its strange that those changes didn't get picked up the first time, I've since pushed a commit with a fix.

File: src/DataCollection.UWP/Controls/DateTimeEditor.xaml (hunk 1 above)
Comment (nCastle1) [reply src/DataCollection.UWP/Controls/DateTimeEditor.xaml:47]: I'm leaving as-is, only because that name wasn't introduced in this PR.

File: src/DataCollection.UWP/LocalizedStrings/Resources.resw [hunk 3]
```diff
@@ -540,4 +540,34 @@
...
+    <value>mm/dd/yyyy</value>
+  </data>
+  <data name="DTE_ResetButton_Tooltip" xml:space="preserve">
+    <value>Clear date and time</value>
+  </data>
+  <data name="DTE_TimeButton_Tooltip" xml:space="preserve">
+    <value>Select a time</value>
+  </data>
+  <data name="DTE_TimePart_Placeholder" xml:space="preserve">
+    <value>hh:mm AM/PM</value>
```
Comment (esreli) [inline src/DataCollection.UWP/LocalizedStrings/Resources.resw:556]: is uppercase "AM" or "PM" required here?

File: src/DataCollection.WPF_NetFramework/LocalizedStrings/Resources.resx [hunk 4]
```diff
@@ -540,4 +540,34 @@
...
+    <value>Select a time</value>
+  </data>
+  <data name="DTE_TimePart_Placeholder" xml:space="preserve">
+    <value>hh:mm AM/PM</value>
+  </data>
+  <data name="DTE_ValidationError_Message" xml:space="preserve">
+    <value>Set date and time to valid values before moving on</value>
+  </data>
+  <data name="DTE_EmptyDate_Message" xml:space="preserve">
+    <value>Saved as: null (no date/time)</value>
```
Comment (esreli) [inline src/DataCollection.WPF_NetFramework/LocalizedStrings/Resources.resx]: "Null" is developer terminology way more-so then it is end-user terminology. Can we nix the 'null'? I'd understand if this was a Calcite team request, it just seems out of place here.

```suggestion
    <value>Saved as: (no date/time)</value>
```

**Review by esreli — COMMENTED** [review]
I didn't test the tool and am trusting you and Zach having do so. The demonstration was quite nice and I trust the edge cases you were working on have been accounted for.

I have just two pedantic comments otherwise this LGTM.

File: src/DataCollection.UWP/LocalizedStrings/Resources.resw [hunk 5]
```diff
@@ -540,4 +540,34 @@
...
+  </data>
+  <data name="DTE_ValidationError_Message" xml:space="preserve">
+    <value>Set date and time to valid values before moving on</value>
+  </data>
+  <data name="DTE_EmptyDate_Message" xml:space="preserve">
+    <value>Saved as: null (no date/time)</value>
```
Comment (esreli) [inline src/DataCollection.UWP/LocalizedStrings/Resources.resw]:
```suggestion
    <value>Saved as: (no date/time)</value>
```

File: src/DataCollection.UWP/LocalizedStrings/Resources.resw (hunk 3 above)
Comment (nCastle1) [reply src/DataCollection.UWP/LocalizedStrings/Resources.resw:556]: If the user enters a time in 24-hour format or omits AM/PM, it will be converted to 12 hour AM/PM format. AM/PM isn't necessarily required, but it is clear about the behavior and expectations. Lowercase am/pm also works, but will be shown as capital AM/PM once committed

Comment (esreli) [reply src/DataCollection.UWP/LocalizedStrings/Resources.resw:556]: As OSA needs it, I think this solution is good as-is. I think it would be worth considering if the Calcite team would want locale support baked in to the tool in its initial implementation.

**Review by esreli — APPROVED** [review]
As a component of Data Collection, this tool is really awesome and ready to go 👍
